### PR TITLE
[scheduling] Moved node affinity from annotations to api fields. #35518

### PR DESCRIPTION
**Pull request by rrati**
Converted node affinity from annotations to api fields

Fixes: #35518 
Related: #25319
Related: #34508

**Release note**:
```release-note
Node affinity has moved from annotations to api fields in the pod spec.  Node affinity that is defined in the annotations will be ignored.
```

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37299)

### CODE REVIEW

**Comment by davidopp** [comment]
I didn't look at the PR yet but I'm adding Daniel and Tim as reviewers because we need to make sure this uses the mechanism we're going to use in general for moving stuff to beta.

**Comment by timothysc** [comment]
2 things 
 
1 - needs release note 
2 - breakup commits. 

/cc @derekwaynecarr @aveshagarwal

**Comment by rrati** [comment]
@timothysc Updates completed

File: pkg/api/types.go
Comment (timothysc) [inline pkg/api/types.go]: @smarterclayton @lavalamp What is the official decoration needed?  
xref: #34508

File: pkg/api/v1/types.go
Comment (timothysc) [inline pkg/api/v1/types.go]: same comment re: decoration

File: pkg/client/record/event.go
Comment (timothysc) [inline pkg/client/record/event.go]: seems unrelated, no need.

**Review by timothysc — COMMENTED** [review]
Just did a quick pass.  We need buy in from @kubernetes/sig-api-machinery on the blessed decoration for alpha/beta fields.

File: pkg/api/types.go
Comment (smarterclayton) [inline pkg/api/types.go]: omitempty and +optional, and if you need to support empty vs nil arrays or maps you should not set omitempty and instead use the same construct that ExtraValues in authorization uses.

**Comment by davidopp** [comment]
Not to sound like a broken record but can you check with @lavalamp or @bgrant0607 or @thockin to make sure this conforms to #34508 ? (I'm not saying this because I think it doesn't -- in fact I haven't even looked at the PR. But I think it would be good to check with them.)

**Comment by smarterclayton** [comment]
The only take away that applies here is:

1. Add fields
2. During decoding (preferable) continue to honor annotations as higher priority than the field.

If you can't do it during decoding you'll need to do it in all client code and in server conversions.

**Comment by rrati** [comment]
@smarterclayton Are you suggesting that we should try to be backwards compatible with the annotations implementation?  If so then I disagree.  This is an alpha feature that has not moved to beta.  Providing backwards compatibility would needlessly complicate the code in order to maintain an undesirable alpha implementation.  I think a clean break from the annotations implementation is best.  No one should be relying upon an alpha implementation since it can change.

**Comment by smarterclayton** [comment]
It's questionable whether alpha -> GA requires back compat.  beta -> GA
definitely does.

**Comment by rrati** [comment]
@smarterclayton I agree beta->GA would likely require backwards compat and there could be cases where alpha->GA would need it as well.  In this case though, I don't think we should attempt backwards compat.

**Comment by timothysc** [comment]
From sig-scheduling mtg on Monday, @davidopp would like @lavalamp to once...

**Comment by lavalamp** [comment]
I have no objections. I do recommend your update code should be tolerant to old clients dropping this field when they do updates, or at least you should think about that case.

File: pkg/api/validation/validation.go
Comment (timothysc) [inline pkg/api/validation/validation.go]: you can probably do the variable swap trick on the if line: 

if na := affinity.NodeAffinity; na != nil {

File: plugin/pkg/scheduler/algorithm/priorities/metadata.go
Comment (timothysc) [inline plugin/pkg/scheduler/algorithm/priorities/metadata.go]: This gets converted in your next PR correct?

File: plugin/pkg/scheduler/algorithm/priorities/node_affinity_test.go
Comment (timothysc) [inline plugin/pkg/scheduler/algorithm/priorities/node_affinity_test.go]: Is there an example where this is actually a string array vs. just a string?  Just wondering if we should rationalize the field types.

**Review by timothysc — CHANGES_REQUESTED** [review]
Only minor comments to address.  I would like @davidopp to once over as well, but 90% of this PR is just translation so lgtm for the most part.

File: pkg/api/validation/validation.go
Comment (rrati) [inline pkg/api/validation/validation.go]: Done

File: plugin/pkg/scheduler/algorithm/priorities/metadata.go
Comment (rrati) [inline plugin/pkg/scheduler/algorithm/priorities/metadata.go]: Once both annotation conversations are in that function gets removed.

File: plugin/pkg/scheduler/algorithm/priorities/node_affinity_test.go
Comment (rrati) [inline plugin/pkg/scheduler/algorithm/priorities/node_affinity_test.go]: Yes.  Look at node affinity in scheduler_predicates.go

**Comment by rrati** [comment]
@k8s-bot test this #18620

File: pkg/api/types.go
Comment (davidopp) [inline pkg/api/types.go]: "Scheduling preferences" is too vague. Also preferences doesn't imply that it can be hard constraints or soft preferences.

**Comment by davidopp** [comment]
IIUC you removed the tests for the annotation version, but didn't remove the code for it. Was that intentional? What is the upgrade plan here? This was one of the things I think we were debating in the discussion about how to move things from annotations to fields (i.e. what do you do when you have both, or one but then upgrade/downgrade, etc.).

**Comment by rrati** [comment]
@davidopp All the code for dealing with node affinity in annotations should have been removed.  The annotations based affinity is there for the pod affinity and anti-affinity definitions, which will be handled in a separate PR.

There is no upgrade plan and imo there shouldn't be one.  This was an alpha feature.  Maintaining backwards compatibility increases code complexity and opens up potential issue for no solid reason.  I think a clean break with the original implementation is best.  Slight more heartache for anyone who used the feature, but they also should expect that with an alpha feature.

File: pkg/api/types.go
Comment (rrati) [inline pkg/api/types.go]: What term would you prefer instead?  maybe say constraints and preferences?

**Review by timothysc — APPROVED** [review]

**Comment by timothysc** [comment]
> There is no upgrade plan and imo there shouldn't be one. This was an alpha feature. Maintaining backwards compatibility increases code complexity and opens up potential issue for no solid reason. I think a clean break with the original implementation is best. Slight more heartache for anyone who used the feature, but they also should expect that with an alpha feature. 

@davidopp We evaluated node-affinity via annotations, and found it to be a UX-pain.  It was easy to make a single character mistake and be lost in, "Why is this not working?".  As a result, we surmised:  

1. It's alpha, so no guarantees  
2. It's not strongly typed, which can lead to UX pain scenarios
3. We don't want to maintain it into beta 
... 

Therefore, we should pay the breaking price on the move to beta, and not be stuck with supporting annotations all the way through.  

As @rrati mentioned the existing annotation code is for inter-pod affinity, and he has a follow on PR to convert that as well.

File: pkg/api/types.go
Comment (timothysc) [inline pkg/api/types.go]: +1 to constraints.

Comment (timothysc) [inline pkg/api/types.go]: @davidopp do you have any opinion on comment?

**Comment by davidopp** [comment]
@wojtek-t

File: pkg/api/v1/types_swagger_doc_generated.go
Comment (davidopp) [inline pkg/api/v1/types_swagger_doc_generated.go]: I'm not sure how this ended up as "preferences" but it should be "constraints" (mostly I just want the terminology to be consistent)

**Comment by davidopp** [comment]
I just had one minor comment, otherwise LGTM, so feel free to LGTM once you fix that. If @wojtek-t has a chance to review this today to double-check the PR, that would be great, but if not, I think it is OK to self-LGTM it since you mentioned you are going to be OOO soon and I don't see a good reason to hold this up over the holidays.

**Comment by davidopp** [comment]
Oh and thanks a lot for doing this!

**Comment by davidopp** [comment]
Should we label the field name as "beta" somehow?

**Comment by timothysc** [comment]
> Should we label the field name as "beta" somehow? (to indicate the feature is beta) 

This is where "policy" comes in, and api-machinery folks said it's ok.  

Should that change, we'll happily change, but folks "**really need to write down the policy**" and update the docs...

**Pull request merged**
